src/bin/shard/login.rs

```diff
@@ -449,3 +451,228 @@ pub fn buddy_menuchat_succ(clients: &mut ClientMap, state: &mut ShardServerState
 
     Ok(())
 }
+
+pub fn buddy_warp_validation(
+    clients: &mut ClientMap,
+    state: &mut ShardServerState,
+) -> FFResult<()> {
+    let login_server = clients.get_login_server().unwrap();
+    let pkt: sP_LS2FE_REQ_BUDDY_WARP_VALIDATION =
+        *login_server.get_packet(P_LS2FE_REQ_BUDDY_WARP_VALIDATION)?;
+
+    let requester_pc_id = pkt.iWarpRequesterPCID;
+    let requester_pc_uid = pkt.iWarpRequesterPCUID;
+    let receiver_pc_uid = pkt.iWarpReceiverPCUID;
+    let requester_payzone_flag = pkt.iWarpRequesterPayzoneFlag != 0;
+    let requester_shard_id = pkt.iWarpRequesterShardID;
+
+    let mut invalid_warp = |error_code: i32, msg: String| -> FFResult<()> {
+        let response = sP_FE2LS_REP_BUDDY_WARP_VALIDATION_FAIL {
+            iWarpRequesterPCID: requester_pc_id,
+            iWarpReceiverPCUID: receiver_pc_uid,
+            iErrorCode: error_code,
+            iWarpRequesterShardID: requester_shard_id,
+        };
+        log_if_failed(
+            clients
+                .get_login_server()
+                .unwrap()
+                .send_packet(P_FE2LS_REP_BUDDY_WARP_VALIDATION_FAIL, &response),
+        );
+        Err(FFError::build(Severity::Info, msg))
+    };
+
+    // Find the receiver (buddy) player
+    let search = PlayerSearchQuery::ByUID(receiver_pc_uid);
+    let res = search.execute(state);
+    if res.is_none() {
+        return invalid_warp(
```

Comment:
I would definitely log a warning here. If the login server thinks a player is in this shard but they're not, then something has gone wrong in the player tracking system